Amplification limit in recovery (#1862)
quicwg/base-drafts: draft-ietf-quic-recovery.md

When an acknowledgement is received, a new RTT is computed and the timer
SHOULD be set for twice the newly computed smoothed RTT.

When crypto packets are sent, the sender MUST set a timer for the crypto

> timeout period.  Upon timeout, the sender MUST retransmit all unacknowledged

Reviewer: ```suggestion
retransmission interval.  Upon timeout, the sender MUST retransmit all unacknowledged
```

Author: This doc uses timeout and timeout period fairly consistently, so I'd like to keep it as is.

CRYPTO data if possible.

Until the server has validated the client's address on the path, the number of
bytes it can send is limited, as specified in {{QUIC-TRANSPORT}}.

> If not all unacknowledged CRYPTO data can be sent, then all unacknowledged

Reviewer: I think that you mean to say instead

> If all CRYPTO data is acknowledged, and there is nothing for the client to send, then it sends another Initial packet.

What that packet contains clearly doesn't matter (though an ACK would be perfectly OK, as would repeating information that the server already has), so I don't think we need to stipulate that explicitly.  Otherwise, this is a little ambiguous - and maybe in conflict with what we did in #1843 and the next paragraph.

Author: This paragraph talks about the server's behavior, not the client's?

CRYPTO data sent in Initial packets should be retransmitted.  If no bytes

> may be sent, then no alarm should be armed until bytes have been received from

Reviewer: ```suggestion
can be sent, then no alarm should be armed until bytes have been received from
```

the client.

> Because the server may be blocked until more packets are received, the

Reviewer: ```suggestion
Because the server could be blocked until more packets are received, the
```

When an acknowledgement is received, a new RTT is computed and the timer
SHOULD be set for twice the newly computed smoothed RTT.

When crypto packets are sent, the sender MUST set a timer for the crypto
timeout period.  Upon timeout, the sender MUST retransmit all unacknowledged
CRYPTO data if possible.

Until the server has validated the client's address on the path, the number of
bytes it can send is limited, as specified in {{QUIC-TRANSPORT}}.
If not all unacknowledged CRYPTO data can be sent, then all unacknowledged
CRYPTO data sent in Initial packets should be retransmitted.  If no bytes
may be sent, then no alarm should be armed until bytes have been received from
the client.

Because the server may be blocked until more packets are received, the

> client MUST arm the crypto retransmission alarm even if there is no

Reviewer: timer, not alarm
```suggestion
client MUST start the crypto retransmission timer even if there is no
```

When crypto packets are sent, the sender SHOULD set a timer for the crypto
timeout period.  Upon timeout, the sender MUST retransmit all unacknowledged
CRYPTO data if possible.

> Until the server has confirmed the path, it must not send more than 3

Reviewer: rather than "confirmed the path", we should be more consistent and say "the server has validated the client's address on the path"

```suggestion
Until the server has validated the client's address on the path, it must not send more than 3
```

> times the number of received bytes, as specified in the {{QUIC-TRANSPORT}}.

Reviewer: Is this MUST NOT?  You might avoid duplicating the requirement by saying that the server is limited in the amount of data it can send, as specified in {{QUIC-TRANSPORT}}.

> If not all unacknowledged CRYPTO data can be sent, then all CRYPTO data sent

Reviewer 2: "then all unacknowledged CRYPTO data sent"

> Initial encryption should be retransmitted.  If no bytes may be sent,

Author: Hmm, probably Initial packets.

then no alarm should be armed until bytes have been received by the peer.

On each consecutive expiration of the crypto timer without receiving an
acknowledgement for a new packet, the sender SHOULD double the crypto

> handshake timeout and set a timer for this period.

Reviewer: The timeouts are still inconsistently named.  We have "crypto handshake timeout" here, but "crypto timeout" on 321/322.  I think that "crypto retransmission timeout " is most correct.  You should fix the section title as well.

and timer events further below.  The function SetLossDetectionTimer defined
below shows how the single timer is set.

> 

Reviewer: Did you mean to delete all of this?

Author: Yes, I believe it's duplicative of the more detailed text above.

Reviewer: OK.  It seemed entirely unrelated, but I see we have text on both elsewhere.
